### ISSUE TITLE
Wildcard properties

### ISSUE THREAD

**Issue by robertbbb** [eclipse/rdf4j#2078]
Hi, 
There are a number of tutorials and solutions that make use of "wildcard properties" in order to identify paths in a graph
(see https://stackoverflow.com/questions/30916040/sparql-is-there-any-path-between-two-nodes)

Such wildcards typically look like
(<>|!<>)|^(<>|!<>)
or
(:|!:)|^(:|!:)

The support in RDF4J appears to be limited or unpredictable - for example:
- the following works to answer if a directed path exists between X and Y:
ask { :X  (:|!:)* :Y }
- however, the following does not detect a non-directed path (althought it's supposed to look for edges regardless of their direction):
ask { :X  ((:|!:)|^(:|!:))* :Y }

Can this be made consistent, or is there a good reason why only the first version works?

Thanks,
Rob

**Comment by hmottestad** [comment]
Could you provide a testfile and a testquery were `{ :X (:|!:)* :Y }` works but `{ :X ((:|!:)|^(:|!:))* :Y }` doesn't?

**Comment by robertbbb** [comment]
@prefix : <http://example.org/>.
:Mary :parentOf :Jim.
:Jim :knows :Jane.
:Jane :worksFor :IBM.

The following works as expected - it finds a directec path from Mary to IBM
ask where
{
  :Mary (:|!:)* :IBM
}

When we reverse the order, the following should work but returns No: 
ask where
ask {
  :IBM ((:|!:)|^(:|!:))* :Mary
}

Also the following responds with No:
ask { :IBM (^(:|!:))* :Mary}

**Comment by hmottestad** [comment]
I've simplified it as follows:

```
select * where {
    :IBM ^(:|!:) ?a.
}
```

Query plan is:
```
01 Projection
02   ProjectionElemList
03      ProjectionElem "a"
04   Union
05      StatementPattern
06         Var (name=a)
07         Var (name=_const_7193c6e1_uri, value=http://example.org/, anonymous)
08         Var (name=_const_1bf5dbd3_uri, value=http://example.org/IBM, anonymous)
09      Filter
10         Compare (!=)
11            Var (name=_anon_5babb2a0_8809_4618_bf71_4ec4d72253d5, anonymous)
12            ValueConstant (value=http://example.org/)
13         StatementPattern
14            Var (name=_const_1bf5dbd3_uri, value=http://example.org/IBM, anonymous)
15            Var (name=_anon_5babb2a0_8809_4618_bf71_4ec4d72253d5, anonymous)
16            Var (name=a)
```

Issue is that statement pattern on line 13-16 should be reversed. Should be:

```
13         StatementPattern
14            Var (name=a)
15            Var (name=_anon_5babb2a0_8809_4618_bf71_4ec4d72253d5, anonymous)
16            Var (name=_const_1bf5dbd3_uri, value=http://example.org/IBM, anonymous)
```

**Comment by hmottestad** [comment]
`:IBM ((:|!:)|(^:|!^:))* ?a` where I have pushed `^` into the path seems to work fine. However, if I specify `:Jane` instead of `?a` I get:

 `Exception in thread "main" java.lang.ClassCastException: class org.eclipse.rdf4j.query.algebra.ValueConstant cannot be cast to class org.eclipse.rdf4j.query.algebra.Var (org.eclipse.rdf4j.query.algebra.ValueConstant and org.eclipse.rdf4j.query.algebra.Var are in unnamed module of loader 'app')`

**Comment by hmottestad** [comment]
We now have a branch with some tests for this issue: https://github.com/eclipse/rdf4j/pull/2080

The ClassCast exception has been fixed, so rewriting the query by pushing the inversion works. The reported issue has not been fixed. I'm at a loss trying to figure out what the code should be doing...help would be nice.

Issue is somewhere in this method: https://github.com/eclipse/rdf4j/blob/474c966a6de74469a9dc0d874e60a25e3fa1395f/core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/TupleExprBuilder.java#L1253

**Comment by hmottestad** [comment]
This has now been fixed and will be released in either 3.1.4 or the next 3.2 milestone release.
